src/services/utils.ts

```diff
@@ -7,71 +7,109 @@
 
 import { getUserToken } from '../redux/commonStore';
 
-export const backendFetch = (url: string, init: any, token?: string) => {
+export type InitRequest = Partial<RequestInit>;
+export type Token = string;
+
+export interface ErrorWithStatus extends Error {
+    status?: number;
+}
+
+export function backendFetch<T>(
+    url: string,
+    init: any,
+    token?: string
+): Promise<T> {
     const initCopy = prepareRequest(init, token);
     return safeFetch(url, initCopy);
-};
+}
 
-export const backendFetchJson = (url: string, init: any, token?: string) => {
-    const initCopy = prepareRequest(init, token);
-    return safeFetch(url, initCopy).then((safeResponse) =>
-        safeResponse.status === 204 ? null : safeResponse.json()
+export async function backendFetchJson(
+    url: string,
+    init?: any,
+    token?: string
+) {
+    return (await backendFetch<Response>(url, init, token)).json();
+}
+
+export async function backendFetchText(
+    url: string,
+    init?: any,
+    token?: string
+) {
+    return (await backendFetch<Response>(url, init, token)).text();
+}
+
+export async function backendFetchFile(
+    url: string,
+    init?: any,
+    token?: string
+) {
+    return (await backendFetch<Response>(url, init, token)).blob();
+}
+
+export enum FileType {
+    ZIP = 'ZIP',
+}
+
+export const downloadFile = (blob: Blob, filename: string, type?: FileType) => {
+    let contentType;
+    if (type === FileType.ZIP) {
+        contentType = 'application/octet-stream';
+    }
+    const href = window.URL.createObjectURL(
+        new Blob([blob], { type: contentType })
     );
+    const link = document.createElement('a');
+    link.href = href;
+    link.setAttribute('download', filename);
+    document.body.appendChild(link);
+    link.click();
+    document.body.removeChild(link);
 };
 
-const prepareRequest = (init: any, token?: string) => {
+function prepareRequest(init?: InitRequest, token?: Token): RequestInit {
     if (!(typeof init === 'undefined' || typeof init === 'object')) {
         throw new TypeError(
-            'First argument of prepareRequest is not an object : ' + typeof init
+            `Argument 2 of backendFetch is not an object ${typeof init}`
         );
     }
-    const initCopy = Object.assign({}, init);
+    const initCopy: RequestInit = Object.assign({}, init);
     initCopy.headers = new Headers(initCopy.headers || {});
     const tokenCopy = token ?? getUserToken();
     initCopy.headers.append('Authorization', 'Bearer ' + tokenCopy);
     return initCopy;
-};
+}
 
-const safeFetch = (url: string, initCopy: any) => {
+function safeFetch<T>(url: string, initCopy: any): Promise<T> {
     return fetch(url, initCopy).then((response) =>
         response.ok ? response : handleError(response)
-    );
-};
+    ) as Promise<T>;
+}
```

Comment:
```suggestion
function safeFetch(url: string, initCopy: RequestInit): Promise<Response> {
    return fetch(url, initCopy).then((response) =>
        response.ok ? response : handleError(response)
    );
}
```
* `initCopy` is of type `RequestInit`
* `safeFetch` always return a `Response` or failed promise

Reply:
Done